### ISSUE TITLE
(module name): short issue description

### ISSUE THREAD

**Issue by urko-b** [aws/aws-cdk#16686]
I'm using `aws-cdk-go` https://github.com/aws/aws-cdk-go to deploy a stack with some functionalities. As far as I create a bucket resource using `NewBucket` call from `github.com/aws/aws-cdk-go/awscdk/awss3` package I get an error that says the bucket already exists. I mean, I think this should check if it exists and omit creating again.

```
 0/101 |7:54:10 AM | CREATE_FAILED        | AWS::S3::Bucket                   | cdk-wiserentr-stack-dev-user-docs (cdkwiserentrstackdevuserdocsC2F72F08) cdk-wiserentr-stack-dev-user-docs already exists
	new Bucket (/tmp/jsii-kernel-aeePMk/node_modules/monocdk/lib/aws-s3/lib/bucket.js:589:26)
	\_ /tmp/jsii-runtime.481357249/lib/program.js:8367:58
	\_ Kernel._wrapSandboxCode (/tmp/jsii-runtime.481357249/lib/program.js:8795:24)
	\_ Kernel._create (/tmp/jsii-runtime.481357249/lib/program.js:8367:34)
	\_ Kernel.create (/tmp/jsii-runtime.481357249/lib/program.js:8108:29)
	\_ KernelHost.processRequest (/tmp/jsii-runtime.481357249/lib/program.js:9692:36)
	\_ KernelHost.run (/tmp/jsii-runtime.481357249/lib/program.js:9655:22)
	\_ Immediate._onImmediate (/tmp/jsii-runtime.481357249/lib/program.js:9656:46)
	\_ processImmediate (internal/timers.js:464:21)
```

### Environment

  - **CDK CLI Version  :**
  - **Framework Version:**
  - **Node.js Version:** 
  - **OS               :**
  - **Language (Version):** 

### Other

--- 

This is :bug: Bug Report

**Comment by peterwoodworth** [comment]
Had you originally created the bucket in your CDK project? And if so, did you modify the bucket in any way?

**Comment by urko-b** [comment]
Yeah I have my stack and I'm creating using a pipeline with `cdk deploy`

```

func NewBucket(stack constructs.Construct, name string) awss3.Bucket {
	return awss3.NewBucket(stack, jsii.String(name), &awss3.BucketProps{
		BucketName: jsii.String(name),
	})
}


-----
	docsBucketName := fmt.Sprintf("%s-user-docs", name)
	s3.NewBucket(appStack, docsBucketName)
```

**Comment by urko-b** [comment]
Is strange because I do the same for aws event bus and It does not try create it again when It already exists

**Comment by peterwoodworth** [comment]
My guess on what's happening here is that you changed the [`logical id`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resources-section-structure.html) of the bucket. Changing the logical ID of a resource in your code will cause CloudFormation to [replace](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-update-behaviors.html#update-replacement) the resource. When replacing a resource, CloudFormation will first create a copy of the resource, then delete the old one. In the case of an S3 bucket, this would cause a failure if you've specified the physical bucketName in your code because CloudFormation would be trying to create a bucket with a name that's already taken!

**Comment by urko-b** [comment]
Hi @peterwoodworth But I mean. Why with event-bus creation is not happening the same? I think the NewBucket method should check if the bucket already exists and then do not create it. Otherwise S3 bucket creation using the cdk-stack code is not reusable. I only can run once the NewBucket creation... Is not friendly for developers.

**Comment by urko-b** [comment]
I have the same code for event-bus creation and I do not have any problem with this. I think aws-cdk should 

```


func New(stack constructs.Construct, name string) awsevents.EventBus {
	return awsevents.NewEventBus(stack, jsii.String(name), &awsevents.EventBusProps{
		EventBusName: jsii.String(name),
	})
}
----
eventBus := bus.New(appStack, eventBusName)
```

**Comment by peterwoodworth** [comment]
There seems to be some confusion here

The CDK synthesizes your code into a CloudFormation template. This CloudFormation template is then used to create a stack through CloudFormation. Please read [this page](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-update-behaviors.html) for more information on how updating CloudFormation stacks works.

I want to bring attention to a couple things mentioned in this page:

> Resources that haven't changed run without disruption during the update process.

Like I said above, CloudFormation would not be trying to create an S3 bucket again unless you made a change which triggers a replacement. This leads me to the next thing I want to bring attention to:

> For updated resources, AWS CloudFormation uses one of the following update behaviors:
> Replacement
> AWS CloudFormation recreates the resource during an update, which also generates a new physical ID. AWS CloudFormation usually creates the replacement resource first, changes references from other dependent resources to point to the replacement resource, and then deletes the old resource. For example, if you update the AvailabilityZone property of an AWS::EC2::Instance resource type, AWS CloudFormation creates a new resource and replaces the current EC2 Instance resource with the new one.

Looking at the [CloudFormation page for a Bucket](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket.html#cfn-s3-bucket-name), only two properties will cause a replacement when changed: `BucketName` and `ObjectLockEnabled`. Changing the former will by definition not allow you to run into this issue, and it looks like you don't have the latter property enabled. So, because changing the Logical ID of a resource will _also_ trigger a replacement, I suspect this is what you did.

**Comment by urko-b** [comment]
@peterwoodworth ok so which property should I enable to keep the code as It is? Thanks in advance

**Comment by peterwoodworth** [comment]
So in your code you have

```go
func NewBucket(stack constructs.Construct, name string) awss3.Bucket {
	return awss3.NewBucket(stack, jsii.String(name), &awss3.BucketProps{
		BucketName: jsii.String(name),
	})
}
```

You're specifying the `BucketName` property here. This sets the physical name of the bucket to whatever you want it to be. If you don't fill this property in, a physical name will automatically be generated. [Here is a link to our documentation on physical names](https://docs.aws.amazon.com/cdk/latest/guide/resources.html#resources_physical_names). We generally advise against setting the physical name yourself, because of the following:

> Assigning physical names to resources has some disadvantages in AWS CloudFormation. Most importantly, any changes to deployed resources that require a resource replacement, such as changes to a resource's properties that are immutable after creation, will fail if a resource has a physical name assigned. If you end up in a state like that, the only solution is to delete the AWS CloudFormation stack, then deploy the AWS CDK app again. See the [AWS CloudFormation documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-name.html) for details.

OPTION 1: So, with this knowledge, the easiest way to get around this issue would be to delete your stack and create it again. After doing this, ensure you don't make changes to anything highlighted here: (note: you might not have `ObjectLockEnabled` filled in. You don't have to worry about it if you don't have it filled in)


func NewBucket(stack constructs.Construct, name string) awss3.Bucket {
  return awss3.NewBucket(stack, **jsii.String(name)**, &awss3.BucketProps{
		**BucketName: jsii.String(name)**,
**ObjectLockEnabled: bool**,
	})
}

OPTION 2: Alternatively, if you don't want to delete and recreate your stack, simply don't fill in the `BucketName` property and let the physical name be auto-generated. 

```go
func NewBucket(stack constructs.Construct, name string) awss3.Bucket {
	return awss3.NewBucket(stack, jsii.String(name))
}
```

**Comment by peterwoodworth** [comment]
I'm not sure which part of the first option didn't make sense to you, can you clarify that?

For option 1, *all you need to do* is to destroy and recreate your stack.

**Comment by urko-b** [comment]
Hi @peterwoodworth yes the problem is that that option `ObjectLockEnabled` is not available in `awss3.BucketProps`

**Comment by peterwoodworth** [comment]
@urko-b don't worry about the `ObjectLockEnabled` property. You can ignore it completely. All you have to do is delete and recreate your stack
